[quant][be] Add a check in prepare_qat to make sure the model is in training mode

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70009
* #69959
* __->__ #69879

Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

jerryzh168 committed:
[quant][be] Add a check in prepare_qat to make sure the model is in training mode

Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

supriyar commented on torch/ao/quantization/quantize.py:
```diff
@@ -435,6 +435,7 @@ def prepare_qat(model, mapping=None, inplace=False, allow_list=None):
         allow_list: a set that lists out allowable modules to be propagated with qconfig
     """
     torch._C._log_api_usage_once("quantization_api.quantize.prepare_qat")
+    assert model.training, "prepare_qat only works on models in training mode"
```
Does it make sense to not have this restriction on the model? our APIs should work on the model irrespective of it being in `train`/`eval` mode.

jerryzh168 replied:
we do have this assumption currently, although it's implicit. I think we can relax the constraint later

jerryzh168 replied:
we will need to refactor these to remove the constraint: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fuser_method_mappings.py#L32, will add the PR in this stack

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

jerryzh168 committed:
Update on "[quant][be] Add a check in prepare_qat to make sure the model is in training mode"


Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).
